### PR TITLE
R4R: Handle required fee for creating a gateway

### PR DESCRIPTION
**Pull request by secret2830**
#1471

### CODE REVIEW

File: app/v1/asset/fees.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"fmt"
+
+	bank "github.com/irisnet/irishub/app/v1/bank"
+	sdk "github.com/irisnet/irishub/types"
+)
+
+var (
+	FeeFactorSetBySize = []string{
```
Comment (zhangyelong) [inline app/v1/asset/fees.go]: (ln(symbol_length) / ln3)^4

Comment (secret2830) [reply app/v1/asset/fees.go]: Yes

File: app/v1/asset/msgs.go [hunk 2]
```diff
@@ -118,21 +119,23 @@ func (msg MsgIssueAsset) GetSigners() []sdk.AccAddress {
...
-	Moniker  string         `json:"moniker"`  //  the globally unique name of the gateway
-	Identity string         `json:"identity"` //  the identity of the gateway
-	Details  string         `json:"details"`  //  the description of the gateway
-	Website  string         `json:"website"`  //  the external website of the gateway
+	Owner     sdk.AccAddress `json:"owner"`      //  the owner address of the gateway
+	Moniker   string         `json:"moniker"`    //  the globally unique name of the gateway
+	Identity  string         `json:"identity"`   //  the identity of the gateway
+	Details   string         `json:"details"`    //  the description of the gateway
+	Website   string         `json:"website"`    //  the external website of the gateway
+	CreateFee sdk.Coins      `json:"create_fee"` // the fee for gateway creation
```
Comment (haifengxi) [inline app/v1/asset/msgs.go]: CreateFee --> Fee

File: app/v1/asset/msgs.go [hunk 3]
```diff
@@ -118,21 +119,23 @@ func (msg MsgIssueAsset) GetSigners() []sdk.AccAddress {
...
-		Moniker:  moniker,
-		Identity: identity,
-		Details:  details,
-		Website:  website,
+		Owner:     owner,
+		Moniker:   moniker,
+		Identity:  identity,
+		Details:   details,
+		Website:   website,
+		CreateFee: createFee,
```
Comment (haifengxi) [inline app/v1/asset/msgs.go]: CreateFee --> Fee

File: app/v1/asset/params.go [hunk 4]
```diff
@@ -18,8 +19,8 @@ const (
 var (
 	KeyAssetTaxRate         = []byte("AssetTaxRate")
 	KeyIssueFTBaseFee       = []byte("IssueFTBaseFee")
-	KeyMintFTBaseFeeRatio   = []byte("MintFTBaseFeeRatio")
-	KeyCreateGatewayFee     = []byte("CreateGatewayFee")
+	KeyMintFTFeeRatio       = []byte("MintFTBaseRatio")
+	KeyCreateGatewayBaseFee = []byte("CreateGatewayBaseFee")
```
Comment (haifengxi) [inline app/v1/asset/params.go]: []byte("MintFTFeeRatio")

File: app/v1/asset/msgs.go (hunk 2 above)
Comment (secret2830) [reply app/v1/asset/msgs.go]: Done

File: app/v1/asset/msgs.go (hunk 3 above)
Comment (secret2830) [reply app/v1/asset/msgs.go]: Done

File: app/v1/asset/params.go (hunk 4 above)
Comment (secret2830) [reply app/v1/asset/params.go]: Done

File: app/v1/asset/fees.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	totalFee := sdk.NewDec(int64(gatewayBaseFee)).Quo(calcGatewayFeeFactor(moniker))
+
+	// check if the provided fee is enough
+	if fee.AmountOf(sdk.NativeTokenName).LT(totalFee.TruncateInt()) {
+		return ErrInsufficientFee(k.Codespace(), fmt.Sprintf("insufficient gateway create fee; expected %s, got %s", totalFee.TruncateInt(), fee.AmountOf(sdk.NativeTokenMinDenom)))
+	}
+
+	// compute tax and burned coin
+	communityTax := sdk.NewCoin(sdk.NativeTokenName, assetTaxRate.Mul(totalFee).TruncateInt())
+	burnedCoin := sdk.NewCoin(sdk.NativeTokenName, sdk.NewDec(1).Sub(assetTaxRate).Mul(totalFee).TruncateInt())
```
Comment (chengwenxi) [inline app/v1/asset/fees.go]: burnedCoin := (totalFee - communityTax).TruncateInt()
should better.

File: app/v1/asset/fees.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+	// decrease loosen tokens
+	k.bk.DecreaseLoosenToken(ctx, sdk.Coins{burnedCoin})
+
+	return nil
+}
+
+// calcGatewayFeeFactor computes the fee factor of the given moniker
+func calcGatewayFeeFactor(moniker string) sdk.Dec {
+	len := len(moniker)
```
Comment (chengwenxi) [inline app/v1/asset/fees.go]: Do not use keywords as a variable name.

File: app/v1/asset/fees.go (hunk 5 above)
Comment (secret2830) [reply app/v1/asset/fees.go]: Done

File: app/v1/asset/fees.go (hunk 6 above)
Comment (secret2830) [reply app/v1/asset/fees.go]: Done

File: app/v1/asset/fees.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		return err
+	}
+
+	// burn burnedCoin
+	if _, err := k.bk.BurnCoins(ctx, owner, sdk.Coins{burnedCoin}); err != nil {
+		return err
+	}
+
+	// decrease loosen tokens
+	k.bk.DecreaseLoosenToken(ctx, sdk.Coins{burnedCoin})
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: This call to DecreaseLooseToken() is redundant, already taken care of in BurnCoins().

File: app/v1/asset/fees.go [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+	return nil
+}
+
+// calcFeeFactor computes the fee factor of the given string
+// Note: make sure that the name size is examined first
+func calcFeeFactor(name string) sdk.Dec {
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: We could do a better job by encoding the factor calc into this function, instead of relying on a hardcoded []string

Comment (secret2830) [reply app/v1/asset/fees.go]: Done

File: app/v1/asset/fees.go (hunk 7 above)
Comment (secret2830) [reply app/v1/asset/fees.go]: Done

**Pull request merged**